File: src-tauri/src/terminal_resources.py
"""On-demand Linux terminal resources, loaded into the existing WSL bridge.
No daemon, worker, shared socket or process supervisor. Per-spawn markers are
inherited by descendants; the explicitly marked root is never a stop target.
"""
import os as _tr_os
import signal as _tr_signal
import time as _tr_time
import select as _tr_select
from pathlib import Path as _tr_Path

_TR_PREV = {}


def _tr_marker(value):
    if not isinstance(value, str) or not value or len(value) > 128 or any(
        not (c.isascii() and (c.isalnum() or c in "-_")) for c in value
    ):
        raise ValueError("Invalid terminal generation")
    return value


def _tr_environment(pid):
    try:
        with open(f"/proc/{pid}/environ", "rb") as stream:
            data = stream.read(1024 * 1024 + 1)
        if len(data) > 1024 * 1024:
            raise ValueError("Process environment exceeds resource sampling limit")
        values = dict(field.split(b"=", 1) for field in data.split(b"\0") if b"=" in field)
        marker = values.get(b"MONOCODE_PTY", b"").decode("ascii")
        root = int(values.get(b"MONOCODE_PTY_ROOT", b"0"))
        return marker, root
    except (ProcessLookupError, FileNotFoundError, PermissionError, UnicodeError):
        return None
    except ValueError:
        return None


def _tr_stat(pid):
    try:
        text = _tr_Path(f"/proc/{pid}/stat").read_text()
        end = text.rfind(")")
        fields = text[end + 2:].split()
        return {"name": text[text.find("(") + 1:end], "state": fields[0],
                "started": int(fields[19]), "cpu": int(fields[11]) + int(fields[12]),
                "rss": max(0, int(fields[21])) * _tr_os.sysconf("SC_PAGE_SIZE")}
    except (ProcessLookupError, FileNotFoundError):
        return None
    except (IndexError, ValueError, PermissionError) as error:
        raise ValueError(f"Cannot sample Linux process {pid}: {error}") from error


def _tr_snapshot(markers):
    trees = {marker: {} for marker in markers}
    deadline = _tr_time.monotonic() + 5
    with _tr_os.scandir("/proc") as entries:
        count = 0
        for entry in entries:
            if not entry.name.isdigit():
                continue
            count += 1
            if count > 32768 or _tr_time.monotonic() > deadline:
                raise ValueError("Linux process sampling limit reached; no partial sample was returned")
            pid = int(entry.name)
            env = _tr_environment(pid)
            if not env or env[0] not in trees:
                continue
            parsed = _tr_stat(pid)
            if parsed and parsed["state"] != "Z":
                trees[env[0]][pid] = {**parsed, "root": env[1]}
    return trees


def _tr_root(marker, members):
    roots = {row["root"] for row in members.values()}
    if len(roots) != 1:
        raise ValueError("Linux terminal root is unavailable or ambiguous")
    root = next(iter(roots))
    if root <= 1 or root not in members or _tr_environment(root) != (marker, root):
        raise ValueError("Linux terminal root exited or changed; refresh")
    return root


def _tr_stats(markers):
    trees = _tr_snapshot(markers)
    now = _tr_time.monotonic()
    hz = _tr_os.sysconf("SC_CLK_TCK")
    # Keep baselines from other windows without unbounded retention.
    for key, (_, at) in list(_TR_PREV.items()):
        if now - at >= 60:
            del _TR_PREV[key]
    if len(_TR_PREV) + sum(len(tree) for tree in trees.values()) > 32768:
        _TR_PREV.clear()
    result = {}
    for marker, members in trees.items():
        try:
            root = _tr_root(marker, members)
            cpu, known = {}, True
            for pid, row in members.items():
                key = (pid, row["started"])
                previous = _TR_PREV.get(key)
                _TR_PREV[key] = (row["cpu"], now)
                if previous and 0.001 < now - previous[1] < 60 and row["cpu"] >= previous[0]:
                    cpu[pid] = (row["cpu"] - previous[0]) / hz / (now - previous[1]) * 100
                else:
                    known = False
            workload = [pid for pid in members if pid != root]
            top = max(workload, key=lambda pid: (cpu.get(pid, 0), members[pid]["rss"])) if workload else None
            result[marker] = {"cpuPct": sum(cpu.values()) if known else None,
                              "rssBytes": sum(row["rss"] for row in members.values()),
                              "processes": len(members), "workload": bool(workload),
                              "top": members[top]["name"] if top else None, "error": None}
        except ValueError as error:
            result[marker] = {"cpuPct": None, "rssBytes": None, "processes": None,
                              "workload": False, "top": None, "error": str(error)}
    return result


def _tr_stop(marker):
    if not hasattr(_tr_os, "pidfd_open") or not hasattr(_tr_signal, "pidfd_send_signal"):
        raise ValueError("Safe workload stopping requires WSL 2 and Python 3.9 or newer")
    members = _tr_snapshot([marker])[marker]
    root = _tr_root(marker, members)
    root_start = members[root]["started"]
    handles = []

    def verify_root():
        current = _tr_stat(root)
        if not current or current["started"] != root_start or _tr_environment(root) != (marker, root):
            raise ValueError("Linux terminal identity changed; no further processes were stopped")

    try:
        for pid, row in members.items():
            if pid <= 1 or pid in (root, _tr_os.getpid()):
                continue
            verify_root()
            try:
                descriptor = _tr_os.pidfd_open(pid)
            except ProcessLookupError:
                continue
            # Append immediately so every failure still closes this handle.
            handles.append(descriptor)
            current = _tr_stat(pid)
            if not current or current["started"] != row["started"] or _tr_environment(pid) != (marker, root):
                handles.pop()
                _tr_os.close(descriptor)
                continue
            try:
                _tr_signal.pidfd_send_signal(descriptor, _tr_signal.SIGTERM)
            except ProcessLookupError:
                pass
        poll = _tr_select.poll()
        for descriptor in handles:
            poll.register(descriptor, _tr_select.POLLIN)
        remaining = set(handles)
        deadline = _tr_time.monotonic() + 1
        while remaining and _tr_time.monotonic() < deadline:
            verify_root()
            for descriptor, _events in poll.poll(50):
                remaining.discard(descriptor)
                poll.unregister(descriptor)
        for descriptor in remaining:
            verify_root()
            try:
                _tr_signal.pidfd_send_signal(descriptor, _tr_signal.SIGKILL)
            except ProcessLookupError:
                pass
    finally:
        for descriptor in handles:
            _tr_os.close(descriptor)
    return None


def terminal_resource_request(request):
    if request["op"] == "terminal_resources":
        markers = request.get("markers")
        if not isinstance(markers, list) or len(markers) > 1024:
            raise ValueError("Invalid terminal resource request")
        return _tr_stats(list(dict.fromkeys(_tr_marker(marker) for marker in markers)))
    if request["op"] == "terminal_stop_workload":
        return _tr_stop(_tr_marker(request.get("marker")))
    raise ValueError("Unknown terminal resource operation")

File: src-tauri/src/terminal_resources.rs
//! Resource-manager adapter over upstream-owned PTYs. No independent process
//! registry, polling thread or terminal lifecycle; only explicit UI requests.
use super::{LivePty, PtyHost};
use serde::Serialize;
use std::collections::HashSet;
use std::sync::Arc;
use tauri::State;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct PtyResource {
    id: String,
    generation: Option<String>,
    alive: bool,
    host: &'static str,
    distro: Option<String>,
    cpu_pct: Option<f32>,
    rss_bytes: Option<u64>,
    processes: Option<u32>,
    workload: bool,
    top: Option<String>,
    error: Option<String>,
}

#[cfg(any(windows, test))]
#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub(crate) struct LinuxResource {
    cpu_pct: Option<f32>,
    rss_bytes: Option<u64>,
    processes: Option<u32>,
    workload: bool,
    top: Option<String>,
    error: Option<String>,
}

#[cfg(windows)]
fn linux_resources(
    live: &[(String, Option<Arc<LivePty>>)],
) -> std::collections::HashMap<String, Result<LinuxResource, String>> {
    let mut locations =
        std::collections::HashMap::<String, (crate::wsl::Location, Vec<String>)>::new();
    for (_, live) in live {
        if let Some(live) = live {
            if let Ok(Some(location)) = crate::wsl::path_location(&live.cwd) {
                locations
                    .entry(location.distribution.to_lowercase())
                    .or_insert_with(|| (location, Vec::new()))
                    .1
                    .push(live.generation.clone());
            }
        }
    }
    let mut result = std::collections::HashMap::new();
    for (location, markers) in locations.into_values() {
        match crate::wsl::terminal_resources(&location, &markers) {
            Ok(mut found) => {
                for marker in markers {
                    result.insert(
                        marker.clone(),
                        found
                            .remove(&marker)
                            .ok_or_else(|| "Linux process sample is unavailable".to_string()),
                    );
                }
            }
            Err(error) => {
                for marker in markers {
                    result.insert(marker, Err(error.clone()));
                }
            }
        }
    }
    result
}

fn owned(live: &LivePty, owner: &str, generation: &str) -> bool {
    live.owner == owner && live.generation == generation
}

fn current(host: &PtyHost, id: &str, live: &Arc<LivePty>) -> bool {
    host.get(id).is_some_and(|value| Arc::ptr_eq(&value, live))
}

#[tauri::command(async)]
pub(crate) fn pty_resources(
    window: tauri::Window,
    host: State<'_, PtyHost>,
    ids: Vec<String>,
) -> Result<Vec<PtyResource>, String> {
    if ids.len() > 1024 || ids.iter().any(|id| id.is_empty() || id.len() > 256) {
        return Err("Too many or invalid terminal identities".into());
    }
    let mut seen = HashSet::new();
    let live: Vec<_> = ids
        .into_iter()
        .filter(|id| seen.insert(id.clone()))
        .map(|id| {
            let entry = host.get(&id).filter(|live| live.owner == window.label());
            (id, entry)
        })
        .collect();
    let roots: Vec<_> = live
        .iter()
        .filter_map(|(_, live)| live.as_ref())
        .filter(|live| {
            crate::wsl::path_location(&live.cwd).is_ok_and(|location| location.is_none())
        })
        .map(|live| live.pid)
        .filter(|pid| *pid > 1)
        .collect();
    let native = crate::proc_stats::sample_trees(&roots);
    #[cfg(windows)]
    let mut linux = linux_resources(&live);
    Ok(live
        .into_iter()
        .map(|(id, live)| {
            let live = live.filter(|live| current(&host, &id, live));
            let mut row = PtyResource {
                id,
                generation: live.as_ref().map(|live| live.generation.clone()),
                alive: live.is_some(),
                host: "native",
                distro: None,
                cpu_pct: None,
                rss_bytes: None,
                processes: None,
                workload: false,
                top: None,
                error: None,
            };
            let Some(live) = live else {
                return row;
            };
            match crate::wsl::path_location(&live.cwd) {
                Ok(Some(location)) => {
                    row.host = "wsl";
                    row.distro = Some(location.distribution);
                    // Never present launcher usage as Linux workload usage.
                    #[cfg(windows)]
                    match linux.remove(&live.generation) {
                        Some(Ok(stat)) => {
                            row.cpu_pct = stat.cpu_pct;
                            row.rss_bytes = stat.rss_bytes;
                            row.processes = stat.processes;
                            row.workload = stat.workload;
                            row.top = stat.top;
                            row.error = stat.error;
                        }
                        Some(Err(error)) => row.error = Some(error),
                        None => row.error = Some("Linux process sample is unavailable".into()),
                    }
                    #[cfg(not(windows))]
                    {
                        row.error = Some("WSL terminals require the native Windows app".into());
                    }
                }
                Err(error) => row.error = Some(error),
                Ok(None)
                    if live.process_started.is_none()
                        || crate::proc_stats::identity(live.pid).ok().flatten()
                            != live.process_started =>
                {
                    row.error =
                        Some("Terminal process identity is unavailable or changed; refresh".into());
                }
                Ok(None) => match &native {
                    Ok(stats) => {
                        if let Some(stat) = stats.get(&live.pid) {
                            row.cpu_pct = stat.cpu_pct;
                            row.rss_bytes = stat.rss_bytes;
                            row.processes = Some(stat.processes);
                            row.workload = stat.workload;
                            row.top = stat.top.clone();
                        } else {
                            row.error = Some("Process exited or cannot be sampled; refresh".into());
                        }
                    }
                    Err(error) => row.error = Some(error.clone()),
                },
            }
            row
        })
        .collect())
}

#[tauri::command(async)]
pub(crate) fn pty_kill_workload(
    window: tauri::Window,
    host: State<'_, PtyHost>,
    id: String,
    generation: String,
) -> Result<(), String> {
    let live = host.get(&id).ok_or("Terminal is not running")?;
    if !owned(&live, window.label(), &generation) {
        return Err("Terminal identity changed; refresh before stopping".into());
    }
    if let Some(location) = crate::wsl::path_location(&live.cwd)? {
        if !current(&host, &id, &live) {
            return Err("Terminal identity changed; refresh".into());
        }
        #[cfg(windows)]
        return crate::wsl::terminal_stop_workload(&location, &live.generation);
        #[cfg(not(windows))]
        {
            let _ = location;
            return Err("WSL terminals require the native Windows app".into());
        }
    }
    let started = live
        .process_started
        .ok_or("Terminal process identity is unavailable; nothing was stopped")?;
    crate::proc_stats::stop_workload(live.pid, started, || current(&host, &id, &live))
}

#[cfg(all(test, unix))]
mod tests {
    use super::*;
    use std::sync::Mutex;
    #[test]
    fn ownership_requires_window_and_current_spawn_generation() {
        let host = PtyHost::new();
        let make = |generation: &str| {
            Arc::new(LivePty {
                owner: "window-a".into(),
                generation: generation.into(),
                process_started: None,
                cwd: "/fixture".into(),
                writer: Mutex::new(Box::new(std::io::sink())),
                master_fd: -1,
                pid: 0,
            })
        };
        let first = make("spawn-a");
        host.insert("terminal".into(), first.clone());
        assert!(owned(&first, "window-a", "spawn-a"));
        assert!(!owned(&first, "window-b", "spawn-a"));
        assert!(!owned(&first, "window-a", "spawn-b"));
        assert!(current(&host, "terminal", &first));
        host.insert("terminal".into(), make("spawn-b"));
        assert!(!current(&host, "terminal", &first));
        host.remove("terminal");
    }
}

#[cfg(test)]
mod linux_tests {
    use super::*;
    #[test]
    fn linux_unknown_fields_are_not_zero_measurements() {
        let row: LinuxResource = serde_json::from_str(r#"{"cpuPct":null,"rssBytes":null,"processes":null,"workload":false,"top":null,"error":"unavailable"}"#).unwrap();
        assert_eq!(
            (
                row.cpu_pct,
                row.rss_bytes,
                row.processes,
                row.workload,
                row.top,
                row.error
            ),
            (None, None, None, false, None, Some("unavailable".into()))
        );
    }

    #[test]
    fn linux_generation_root_and_pidfd_stopping_are_checked() {
        let fixture = r#"
from unittest.mock import patch
import types
marker = 'test-generation'
base = {'state':'S', 'started':10, 'cpu':5, 'rss':4096, 'root':10, 'name':'sh'}
members = {10:dict(base), 11:{**base,'started':11,'name':'sleep'}, 12:{**base,'started':12,'name':'node'}}
env = lambda pid: (marker, 10)
with patch.dict(globals(), {'_tr_snapshot':lambda markers:{marker:members}, '_tr_environment':env}):
    first = terminal_resource_request({'op':'terminal_resources','markers':[marker]})[marker]
    assert first['cpuPct'] is None and first['rssBytes'] == 12288 and first['processes'] == 3 and first['workload']
    # Root is explicit, not guessed from the oldest process or lowest PID.
    try:
        _tr_root(marker, {11:members[11]})
        assert False
    except ValueError:
        pass
    signals, closed = [], []
    stat = lambda pid: ({**members[pid], 'started':999} if pid == 12 else members[pid])
    class Poll:
        def __init__(self): self.fds = set()
        def register(self, fd, event): self.fds.add(fd)
        def unregister(self, fd): self.fds.remove(fd)
        def poll(self, delay): return [(fd, 1) for fd in self.fds]
    with patch.dict(globals(), {'_tr_stat':stat}), patch.object(_tr_os, 'pidfd_open', lambda pid:pid+100, create=True), patch.object(_tr_os, 'close', closed.append), patch.object(_tr_signal, 'pidfd_send_signal', lambda fd, sig:signals.append((fd,sig)), create=True), patch.object(_tr_select, 'poll', Poll):
        terminal_resource_request({'op':'terminal_stop_workload','marker':marker})
    assert signals == [(111, _tr_signal.SIGTERM)], signals
    assert sorted(closed) == [111,112], closed
for bad in ['', 'bad=value', 'space here', 'ą', 'a'*129, None]:
    try:
        terminal_resource_request({'op':'terminal_stop_workload','marker':bad})
        assert False
    except ValueError:
        pass
"#;
        let script = format!("{}\n{}", include_str!("terminal_resources.py"), fixture);
        let result = std::process::Command::new("python3")
            .args(["-c", &script])
            .output()
            .unwrap();
        assert!(
            result.status.success(),
            "{}",
            String::from_utf8_lossy(&result.stderr)
        );
    }
}
